File: Chapter 02/Python/2.7.1 Weighted Decision Model/decision_maker.py
from typing import Dict, Tuple, List
from option import Option

def pick_option(options: List[Option], weights: Dict[str, float]) -> Tuple[str, str]:
    """
    Picks the best option based on weighted scores.
    Returns the name of the best option and a rationale string.
    """
    best_option = None
    highest_score = float("-inf")
    details = []

    for opt in options:
        # Calculate the weighted score for this option
        score = sum(opt.scores.get(k, 0) * weights.get(k, 0.0) 
                    for k in weights)
        details.append(f"{opt.name}: {score:.2f}")

        if score > highest_score:
            highest_score = score
            best_option = opt

    rationale = f"Scores: {' | '.join(details)}\n -> Based on weights {weights}, we pick **{best_option.name}**."
    return best_option.name, rationale

File: Chapter 02/JS/2.7.1 Weighted Decision Model/decisionMaker.js
/**
 * Picks the best option based on weighted scores.
 * Returns an object containing the name of the best option and a rationale string.
 */
function pickOption(options, weights) {
    let bestOption = null;
    let highestScore = -Infinity;
    const details = [];

    for (const opt of options) {
        // Calculate the weighted score for this option
        let score = 0;
        for (const [key, weight] of Object.entries(weights)) {
            score += (opt.scores[key] || 0) * weight;
        }
        details.push(`${opt.name}: ${score.toFixed(2)}`);

        if (score > highestScore) {
            highestScore = score;
            bestOption = opt;
        }
    }

    // JSON stringify the weights and clean up the quotes to roughly match the Python output
    const weightsString = JSON.stringify(weights).replace(/"/g, "'");
    const rationale = `Scores: ${details.join(' | ')}\n -> Based on weights ${weightsString}, we pick **${bestOption.name}**.`;
    
    return { bestOption: bestOption.name, rationale };
}

module.exports = { pickOption };